File: backends/basecamp.js
var request = require('request');

exports.dispatch = function (userinfo) {
  request({
    method: 'POST',
    url: 'https://basecamp.com/' + process.env.BASECAMP_ACCOUNT +
    '/api/v1/projects/' + process.env.BASECAMP_PROJECT + '/messages.json',
    json: {
      subject: 'Swag for dude/dudette',
      content: userinfo,
      subscribers: process.env.BASECAMP_SUBSCRIBERS.split(',')
    }, auth: {
      user: process.env.BASECAMP_USER,
      pass: process.env.BASECAMP_PWD
    }, headers: {
      'User-Agent': 'ContriBot',
      'Content-Type': 'application/json'
    }
  }, function(e, r, b) {
    console.log(arguments);
  });
};

File: backends/basecamp.go
package backends

import (
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/elbuo8/contribot/contribot"
	"net/http"
	"strings"
)

type BasecampOptions struct {
	Username    string
	Password    string
	Account     string
	Project     string
	Subscribers []string
	Subject     string
}

func Basecamp(opts *BasecampOptions) contribot.Backend {
	return func(sub *contribot.Submission) {
		postToBasecamp(opts, sub)
	}
}

func postToBasecamp(opts *BasecampOptions, sub *contribot.Submission) {
	urlStr := fmt.Sprintf("https://basecamp.com/%s/api/v1/projects/%s/messages.json'", opts.Account, opts.Project)
	blob := make(map[string]string)
	blob["subject"] = opts.Subject
	blob["subscribers"] = strings.Join(opts.Subscribers, ",")
	blob["content"] = fmt.Sprintf("New Contributor!\nName: %s\nAddress: %s\nEmail: %s\nSize: %s", sub.Name, sub.Address, sub.Email, sub.Size)
	body, _ := json.Marshal(blob)
	req, _ := http.NewRequest("POST", urlStr, bytes.NewReader(body))
	req.Header.Set("User-Agent", "ContriBot")
	req.Header.Set("Content-Type", "application/json")
	req.SetBasicAuth(opts.Username, opts.Password)
	http.DefaultClient.Do(req)
}
